### PR TITLE
First iteration to support letter wilds in file paths

### PR DESCRIPTION
**Pull request by jedi22**
Support paths like this:
```
/b%/b%
```
To get:
```
/bin/bash
```

### CODE REVIEW

File: osquery/filesystem/filesystem.cpp [hunk 1]
```diff
@@ -252,18 +252,49 @@ Status resolveLastPathComponent(const boost::filesystem::path& fs_path,
...
+      results.push_back(fs_path.string());
+      return Status(0, "OK");
+    }
+  }catch(const boost::filesystem::filesystem_error& e){
+    // This should catch permission problems
+    return Status(0, "OK");
+  }
+
+  std::vector<std::string> files;
+  Status stat = listFilesInDirectory(fs_path.parent_path(), files);
```
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: auto

File: osquery/filesystem/filesystem.cpp [hunk 2]
```diff
@@ -252,18 +252,49 @@ Status resolveLastPathComponent(const boost::filesystem::path& fs_path,
...
     return stat;
   }
+
+  // Is the path a file
+  try{
+    if (boost::filesystem::is_regular_file(fs_path)) {
+      results.push_back(fs_path.string());
+      return Status(0, "OK");
+    }
+  }catch(const boost::filesystem::filesystem_error& e){
```
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: clang-format

File: osquery/filesystem/filesystem.cpp [hunk 3]
```diff
@@ -363,7 +408,7 @@ Status resolveFilePattern(std::vector<std::string> components,
 
 Status resolveFilePattern(const boost::filesystem::path& fs_path,
                           std::vector<std::string>& results) {
-  return resolveFilePattern(split(fs_path.string(), "/"), results, 0, 0);
+  return resolveFilePattern(split(fs_path.string(), "/"), results);
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: The whole thing looks complicated but solid. Some of the wildcard checking/tokenizing/path Frankensteining might be clearer if broken out to intermediate methods. 

One note here, looks like you can take a relative path, then on 322, does that join clobber a relative into an absolute?

Comment (jedi22) [inline osquery/filesystem/filesystem.cpp]: Line 322 takes the resolved components and creates the absolute path out of them. The components are the parts of an absolute filepath split at /

File: osquery/filesystem/filesystem_tests.cpp [hunk 4]
```diff
@@ -69,12 +69,38 @@ TEST_F(FilesystemTests, test_wildcard_full_recursion) {
...
+  EXPECT_TRUE(status.ok());
+  EXPECT_NE(std::find(files.begin(), files.end(), "/bin/bash"),
+            files.end());
+  EXPECT_NE(std::find(files.begin(), files.end(), "/bin/sh"),
+            files.end());
+}
+
+TEST_F(FilesystemTests, test_wildcard_three_kinds){
+  std::vector<std::string> files;
+  auto status = resolveFilePattern("/%sr/%/%%", files);
```
Comment (marpaia) [inline osquery/filesystem/filesystem_tests.cpp]: what's the time on these? is it worth doing the /tmp directory tom-foolery again?

Comment (marpaia) [inline osquery/filesystem/filesystem_tests.cpp]: i would assume something like this is pretty fast because you're specifying "sr"

Comment (jedi22) [inline osquery/filesystem/filesystem_tests.cpp]: They can be replaced once we have gmock. This test checks that all wildcards can be chained together exactly as you'd expect them to be able to be. Is there a plan on gmock? or is there another way than both we should consider?

File: osquery/filesystem/filesystem.cpp [hunk 5]
```diff
@@ -343,16 +417,31 @@ Status resolveFilePattern(std::vector<std::string> components,
...
+            return stat;
+          }
+        }
+      }
+      return Status(0, "OK");
     } else {
-      // When we want to support wildcards like "foobar%", this is where it
-      // will happen
+      // There was no special case to be handled, so it's either invalid or
+      // no wildcards.
```
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: you can probably rm this else if there's no unexpected edge cases.

File: osquery/filesystem/filesystem_tests.cpp (hunk 4 above)
Comment (theopolis) [inline osquery/filesystem/filesystem_tests.cpp]: Please please please please please STOP touching the filesystem. Make a structure in /tmp then  replace with a mock later. This is not an acceptable intermediate for mocking.

Comment (marpaia) [inline osquery/filesystem/filesystem_tests.cpp]: +1

**Comment by jedi22** [comment]
```
      Start 20: filesystem_tests
20/31 Test #20: filesystem_tests .................   Passed    0.03 sec
```

**Comment by marpaia** [comment]
Like a boss! Squash, rebase, pass build, merge!

**Pull request merged**